### PR TITLE
boards: x86: add support for UP Squared (Pentium/Celeron)

### PR DESCRIPTION
**Pull request by dcpleung**
This patch set contains the SoC configuration for Apollo Lake, and the board configuration for UP Squared board. This allows Zephyr to run on the UP Squared board.

### CODE REVIEW

File: boards/x86/up_squared/doc/up_squared.rst [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+This board configuration enables kernel support for the `UP Squared`_ board,
+along with the following devices:
+
+* High Precision Event Timer (HPET)
+
+* Serial Ports in Polling and Interrupt Driven Modes
+
+.. note::
+   This board configuration only works with the boards containing
+   Intel |reg| Pentium |trade| SoC, or Intel |reg| Celeton |trade| SoC.
```
Comment (dbkinder) [inline boards/x86/up_squared/doc/up_squared.rst]: ``Celeron`` (misspelled)

File: boards/x86/up_squared/doc/up_squared.rst [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
++===========+============+=======================+
+| HPET      | on-chip    | system clock          |
++-----------+------------+-----------------------+
+| APIC      | on-chip    | interrupt controller  |
++-----------+------------+-----------------------+
+| UART      | on-chip    | serial port-polling;  |
+|           |            | serial port-interrupt |
++-----------+------------+-----------------------+
+
+The kernel currently does not support other hardware features.
```
Comment (dbkinder) [inline boards/x86/up_squared/doc/up_squared.rst]: To clarify,

    The Zephyr kernel currently ...

File: boards/x86/up_squared/doc/up_squared.rst [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+| 2   | HPET    | timer driver             |
++-----+---------+--------------------------+
+| 4   | UART    | serial port when used in |
+|     |         | interrupt mode           |
++-----+---------+--------------------------+
+
+HPET System Clock Support
+-------------------------
+
+Galileo uses HPET timing with legacy-free timer support. The board
```
Comment (dbkinder) [inline boards/x86/up_squared/doc/up_squared.rst]: Replace ``Galileo`` with ``UP Squared``  :)

File: boards/x86/up_squared/doc/up_squared.rst [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+Memory Mappings
+===============
+
+This board configuration uses default hardware memory map
+addresses and sizes.
+
+Programming and Debugging
+*************************
+
+Use the following procedures for booting an image on a Galileo board.
```
Comment (dbkinder) [inline boards/x86/up_squared/doc/up_squared.rst]: Replace ``Galileo`` with ``UP Squared``  :)

File: boards/x86/up_squared/doc/up_squared.rst [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+   :file:`$ZEPHYR_BASE/boards/x86/common/scripts/grub/bin/grub_x86_64.efi`.
+
+Preparing the Boot Device
+=========================
+
+Prepare a USB flash drive to boot the Zephyr application image on
+a UP Squared board.
+
+#. Build a Zephyr application; for instance, to build the ``hello_world``
+   application on Galileo:
```
Comment (dbkinder) [inline boards/x86/up_squared/doc/up_squared.rst]: Replace ``Galileo`` with ``UP Squared``  :)

File: boards/x86/up_squared/doc/up_squared.rst [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+      :board: up_squared
+      :goals: build
+
+   .. note::
+
+      A stripped project image file named :file:`zephyr.strip` is automatically
+      created in the build directory after the application is built. This image
+      has removed debug information from the :file:`zephyr.elf` file.
+
+#. Refer to the UP Squared Serial Console Wiki page for instructions on how to
```
Comment (dbkinder) [inline boards/x86/up_squared/doc/up_squared.rst]: Get the link working with:

    #. Refer to the `UP Squared Serial Console Wiki page
       <https://wiki.up-community.org/Serial_console>`_ for instructions on how to
       connect to a serial console.

File: boards/x86/up_squared/doc/up_squared.rst [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+   connect for serial console.
+
+   `<https://wiki.up-community.org/Serial_console>`_
+
+#. Format the USB flash drive as FAT
```
Comment (dbkinder) [inline boards/x86/up_squared/doc/up_squared.rst]: I'd put in ``code-block``s with the actual commands to use for these steps.

File: boards/x86/up_squared/doc/up_squared.rst [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+Boot the UP Squared board from the boot device using GRUB2 via USB flash drive.
+
+Steps
+-----
+
+1. Insert the prepared boot device (USB flash drive) into the UP Squared board.
+
+2. Connect the board to the host system using the serial cable and
+   configure your host system to watch for serial data.  See
+   `https://wiki.up-community.org/Serial_console`_.
```
Comment (dbkinder) [inline boards/x86/up_squared/doc/up_squared.rst]: Replace this line with just the url:

    https://wiki.up-community.org/Serial_console

File: boards/x86/up_squared/doc/up_squared.rst [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+   GRUB2 starts and a menu shows entries for the items you added
+   to the file :file:`grub.cfg`.
+
+6. Select the image you want to boot and press :guilabel:`Enter`.
+
+   When the boot process completes, you have finished booting the
+   Zephyr application image.
+
+   .. note::
+      If the following messages appear during boot, they can be safely
```
Comment (dbkinder) [inline boards/x86/up_squared/doc/up_squared.rst]: change to:

    You can safely ignore this message if it appears:

File: boards/x86/up_squared/doc/up_squared.rst [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+      ignored.
+
+      .. code-block:: console
+
+         WARNING: no console will be available to OS
+
+References
```
Comment (dbkinder) [inline boards/x86/up_squared/doc/up_squared.rst]: Delete this title (and underline)

**Review by dbkinder — DISMISSED** [review]

**Comment by dcpleung** [comment]
Updated the documentation for comments from @dbkinder .

**Review by tyrantlizard77 — DISMISSED** [review]

File: arch/x86/soc/apollo_lake/Kconfig.defconfig [hunk 11]
```diff
@@ -0,0 +1,49 @@
+#
+# Kconfig - Apollo Lake SoC configuration options
+#
+# Copyright (c) 2014-2015 Wind River Systems, Inc.
```
Comment (ramakrishnapallala) [inline arch/x86/soc/apollo_lake/Kconfig.defconfig]: this is new file anyway, why do we need this copyright?

File: arch/x86/soc/apollo_lake/linker.ld [hunk 12]
```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright (c) 2011-2014, Wind River Systems, Inc.
```
Comment (ramakrishnapallala) [inline arch/x86/soc/apollo_lake/linker.ld]: Intel 2018?

File: arch/x86/soc/apollo_lake/soc.h [hunk 13]
```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright (c) 2010-2015, Wind River Systems, Inc.
```
Comment (ramakrishnapallala) [inline arch/x86/soc/apollo_lake/soc.h]: why this line is needed?

File: boards/x86/up_squared/up_squared.yaml [hunk 14]
```diff
@@ -0,0 +1,11 @@
+identifier: up_squared
+name: UP Squared
+type: mcu
```
Comment (ramakrishnapallala) [inline boards/x86/up_squared/up_squared.yaml]: is mcu is the right name for Apollo lake?

File: boards/x86/up_squared/up_squared_defconfig [hunk 15]
```diff
@@ -0,0 +1,15 @@
+CONFIG_X86=y
+CONFIG_SOC_APOLLO_LAKE=y
+CONFIG_BOARD_UP_SQUARED=y
+CONFIG_CPU_APOLLO_LAKE=y
```
Comment (ramakrishnapallala) [inline boards/x86/up_squared/up_squared_defconfig]: what is the difference b/w CONFIG_CPU_APOLLO_LAKE & CONFIG_SOC_APOLLO_LAKE? i mean can we have any other configuration with CONFIG_CPU_APOLLO_LAKE other than CONFIG_SOC_APOLLO_LAKE?

**Review by ramakrishnapallala — COMMENTED** [review]
minor comments and queries.

File: arch/x86/soc/apollo_lake/Kconfig.soc [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+config SOC_APOLLO_LAKE
+	bool "Intel Apollo Lake Soc"
+	select CPU_APOLLO_LAKE
+	select BOOTLOADER_UNKNOWN
+	select X86_MMU
+	select APPLICATION_MEMORY
```
[the remark this answers is not in the document]
Comment (dcpleung) [reply arch/x86/soc/apollo_lake/Kconfig.soc]: These were copied from the Atom SoC Kconfig. Will remove.

File: arch/x86/soc/apollo_lake/Kconfig.defconfig (hunk 11 above)
Comment (dcpleung) [reply arch/x86/soc/apollo_lake/Kconfig.defconfig]: The file was copied from the arch/x86/soc/atom. So need to preserve copyright.

File: arch/x86/soc/apollo_lake/linker.ld (hunk 12 above)
Comment (dcpleung) [reply arch/x86/soc/apollo_lake/linker.ld]: The file was copied from the arch/x86/soc/atom. Nothing changed so preserve the original copyright.

File: arch/x86/soc/apollo_lake/soc.h (hunk 13 above)
Comment (dcpleung) [reply arch/x86/soc/apollo_lake/soc.h]: The file was copied from the arch/x86/soc/atom. So need to preserve copyright.

File: boards/x86/up_squared/up_squared.yaml (hunk 14 above)
Comment (dcpleung) [reply boards/x86/up_squared/up_squared.yaml]: scripts/sanity_chk/sanitycheck-platform-schema.yaml does not seem to have any other suitable type. Do you want to add one?

File: boards/x86/up_squared/up_squared_defconfig (hunk 15 above)
Comment (dcpleung) [reply boards/x86/up_squared/up_squared_defconfig]: The SOC will select the CPU, so will remove the CPU_* here.

**Comment by dcpleung** [comment]
Removed the lines as indicated by the reviews.

File: arch/x86/soc/apollo_lake/soc.h (hunk 13 above)
Comment (nashif) [reply arch/x86/soc/apollo_lake/soc.h]: You should put your copyright on top if you have made any changes

**Comment by dcpleung** [comment]
Moved the copyright lines.

**Review by nashif — APPROVED** [review]

**Pull request merged**